### PR TITLE
Make memory spill in resource group take effect

### PR DESCRIPTION
**Pull request by guofengrichard**
Resource group memory spill is similiar to statement_mem in
resource queue, the difference is memory spill is calculated
according to the memory quota of the resource group.

The related GUCs, variables and functions shared by both resource
queue and resource group are moved to the namespace resource manager.

### CODE REVIEW

File: src/backend/utils/resscheduler/resqueue.c [hunk 1]
```diff
@@ -2018,3 +2020,172 @@ pg_resqueue_status_kv(PG_FUNCTION_ARGS)
...
+/**
+ * Minimum of two doubles
+ */
+static double minDouble(double a, double b)
+{
+	if (a < b)
+		return a;
+	else
+		return b;
+}
```
Comment (hsyuan) [inline src/backend/utils/resscheduler/resqueue.c]: why don't we use `Min()`?

Comment (pivotal-ning-yu) [inline src/backend/utils/resscheduler/resqueue.c]: @hsyuan this piece of code was moved from src/backend/utils/resscheduler/memquota.c to here, so we didn't touch its logic or implementation. Do you think we should refactor the code in such a case?

Comment (guofengrichard) [inline src/backend/utils/resscheduler/resqueue.c]: These are resource queue codes and we just moved them into a new place.
For this PR, we don't intend to revise resource queue codes.

File: src/backend/utils/resgroup/resgroup.c [hunk 2]
```diff
@@ -1107,6 +1120,7 @@ AssignResGroupOnMaster(void)
 	slot->memLimit = slot->segmentChunks * memoryLimit / 100;
 	slot->memSharedQuota = slot->memLimit * sharedQuota / 100;
 	slot->memQuota = slot->memLimit * (100 - sharedQuota) / concurrency / 100;
+	slot->memSpill = slot->memQuota * spillRatio / 100;
```
Comment (xiong-gang) [inline src/backend/utils/resgroup/resgroup.c]: slot-> memSpill = slot->memLimit * spillRatio / concurrency / 100;

File: src/backend/tcop/pquery.c [hunk 3]
```diff
@@ -672,54 +681,51 @@ PortalStart(Portal portal, ParamListInfo params, Snapshot snapshot,
...
+						if (!SPI_context() || !saveActivePortal || !saveActivePortal->releaseResLock)
+							portal->releaseResLock = ResLockPortal(portal, queryDesc);
+					}
+
+					if (IsResGroupEnabled())
+					{
+						portal->status = PORTAL_QUEUE;
+
+						if (gp_resmanager_memory_policy != RESMANAGER_MEMORY_POLICY_NONE)
+							queryDesc->plannedstmt->query_mem = ResGroupGetQuerySpillMemSize();
```
Comment (xiong-gang) [inline src/backend/tcop/pquery.c]: Is it possible the spill ratio hasn't been correctly set here?

File: src/backend/utils/misc/guc_gp.c [hunk 4]
```diff
@@ -5338,11 +5338,11 @@ struct config_string ConfigureNamesString_gp[] =
 	},
 
 	{
-		{"gp_resqueue_memory_policy", PGC_SUSET, RESOURCES_MGM,
```
Comment (xiong-gang) [inline src/backend/utils/misc/guc_gp.c]: i am a little concern about the compatibility if we rename the existing GUC

File: src/backend/tcop/pquery.c (hunk 3 above)
Comment (xiong-gang) [inline src/backend/tcop/pquery.c]: ResGroupGetQuerySpillMemSize() returns the number of chunks, it doesn't always equal to MB

File: src/test/isolation2/expected/resgroup_memory_hashagg_spill.out [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+DROP
+-- end_ignore
+CREATE ROLE role1_memory_test SUPERUSER;
+CREATE
+CREATE RESOURCE GROUP rg1_memory_test WITH (concurrency=2, cpu_rate_limit=10, memory_limit=10, memory_shared_quota=0, memory_spill_ratio=10);
+CREATE
+SET ROLE TO role1_memory_test;
+SET
+
+set statement_mem="1800";
```
Comment (xiong-gang) [inline src/test/isolation2/expected/resgroup_memory_hashagg_spill.out]: why set statement_mem

File: src/test/isolation2/sql/resgroup_memory_hashagg_spill.sql [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+CREATE ROLE role1_memory_test SUPERUSER;
+CREATE RESOURCE GROUP rg1_memory_test WITH
+(concurrency=2, cpu_rate_limit=10, memory_limit=10, memory_shared_quota=0, memory_spill_ratio=10);
+SET ROLE TO role1_memory_test;
+
+set statement_mem="1800";
+0: ALTER ROLE role1_memory_test RESOURCE GROUP none;
+0: DROP RESOURCE GROUP rg1_memory_test;
+0: CREATE RESOURCE GROUP rg1_memory_test WITH (concurrency=2, cpu_rate_limit=10, memory_limit=30, memory_shared_quota=0, memory_spill_ratio=2);
+0: ALTER ROLE role1_memory_test RESOURCE GROUP rg1_memory_test;
```
Comment (xiong-gang) [inline src/test/isolation2/sql/resgroup_memory_hashagg_spill.sql]: why do we need "0:..." here?

**Comment by xiong-gang** [comment]
About the test cases, can you add some negative cases? for example, when the spill_ratio is set to xxx, the query won't spill.

File: src/backend/utils/resgroup/resgroup.c (hunk 2 above)
Comment (guofengrichard) [inline src/backend/utils/resgroup/resgroup.c]: Yes.

File: src/backend/executor/spi.c [hunk 7]
```diff
@@ -2052,16 +2052,23 @@ static void
...
-				queryDesc->plannedstmt->query_mem =
-					ResourceQueueGetQueryMemoryLimit(queryDesc->plannedstmt,
-													 ActivePortal->queueId);
+				if (IsResQueueEnabled())
+				{
+					queryDesc->plannedstmt->query_mem =
+						ResourceQueueGetQueryMemoryLimit(queryDesc->plannedstmt,
+								ActivePortal->queueId);
+				}
+				else if (IsResGroupEnabled())
```
Comment (pengzhout) [inline src/backend/executor/spi.c]: So in this case, the spill ratio will take no effect when resource group is enabled and it's supper user? i thinks this is not supposed.

File: src/backend/tcop/pquery.c (hunk 3 above)
Comment (guofengrichard) [inline src/backend/tcop/pquery.c]: I do not think it is possible. The spill ratio would be set at the start of every transaction. At this place, transaction should have been started.
I will verify this further.

Comment (guofengrichard) [inline src/backend/tcop/pquery.c]: Actually ResGroupGetQuerySpillMemSize() would return spill size in Bytes. It is calculated as the memSpill in slot left shift by chunkSizeInBits.

File: src/backend/utils/misc/guc_gp.c (hunk 4 above)
Comment (guofengrichard) [inline src/backend/utils/misc/guc_gp.c]: Your concern makes sense. We will keep the existing GUC and introduce into a new GUC 'gp_resgroup_memory_policy'.

File: src/test/isolation2/expected/resgroup_memory_hashagg_spill.out (hunk 5 above)
Comment (guofengrichard) [inline src/test/isolation2/expected/resgroup_memory_hashagg_spill.out]: We neglect to remove 'statement_mem'... Thanks for pointing it out.

File: src/test/isolation2/sql/resgroup_memory_hashagg_spill.sql (hunk 6 above)
Comment (guofengrichard) [inline src/test/isolation2/sql/resgroup_memory_hashagg_spill.sql]: No need of '0:'. Will remove them.

File: src/backend/executor/spi.c (hunk 7 above)
Comment (guofengrichard) [inline src/backend/executor/spi.c]: Fixed this. In the case of resource group enabled and being super user, the spill ratio in resource group will take effect.

File: src/backend/executor/spi.c [hunk 8]
```diff
@@ -2092,62 +2099,69 @@ _SPI_pquery(QueryDesc *queryDesc, bool fire_triggers, long tcount)
...
+							Assert(ActivePortal->releaseResLock);
+
+							_SPI_assign_query_mem(queryDesc);
+						}
 					}
 				}
+				else if (IsResGroupEnabled())
+				{
+					_SPI_assign_query_mem(queryDesc);
+				}
```
Comment (pivotal-ning-yu) [inline src/backend/executor/spi.c]: If we move this resgroup part before the resqueue part and append a `break;` after `_SPI_assign_query_mem()` then we don't have to touch the resqueue code. e.g.:

```C
if (IsResGroupEnabled())
{
	_SPI_assign_query_mem(queryDesc);
	break;
}

if (ActivePortal)
{
...
}
```

From my point of view we should avoid unnecessary changes to existing code.

File: src/backend/utils/resource_manager/memquota.c [hunk 9]
```diff
@@ -97,26 +94,25 @@ typedef struct PolicyEagerFreeContext
 	PlannedStmt *plannedStmt; /* pointer to the planned statement */
 } PolicyEagerFreeContext;
 
-/**
- * GUCs
+/*
```
Comment (pivotal-ning-yu) [inline src/backend/utils/resource_manager/memquota.c]: Why change `/**` to `/*`? The former is the doxygen style mark for inline document, it's not a typo.

File: src/test/isolation2/sql/enable_resgroup.sql [hunk 10]
```diff
@@ -7,8 +7,18 @@
...
+! echo 'overcommit=$(cat /proc/sys/vm/overcommit_ratio)' >>/tmp/.resgroup_mem_helper.sh;
+! echo 'total=$((swap + mem * overcommit / 100))' >>/tmp/.resgroup_mem_helper.sh;
+! echo 'limit=$(((2 << 30) * 10000 * 4 / 3 / total))' >>/tmp/.resgroup_mem_helper.sh;
+! echo '[ "$limit" -le 9000 ] || limit=9000' >>/tmp/.resgroup_mem_helper.sh;
+! echo 'limitf=$(printf "0.%04d" $limit)' >>/tmp/.resgroup_mem_helper.sh;
+! echo 'gpconfig -c gp_resource_group_memory_limit -v $limitf' >>/tmp/.resgroup_mem_helper.sh;
+! bash /tmp/.resgroup_mem_helper.sh;
+! rm -f /tmp/.resgroup_mem_helper.sh;
 ! gpconfig -c gp_resource_manager -v group;
-! gpconfig -c gp_resource_group_cpu_limit -v 0.5;
```
Comment (pivotal-ning-yu) [inline src/test/isolation2/sql/enable_resgroup.sql]: The cpu tests require this GUC to be set to 0.9, so instead of remove this incorrect set to 0.5, please explicitly set it to 0.9 here.

File: src/backend/commands/explain.c [hunk 11]
```diff
@@ -1447,7 +1453,8 @@ explain_outNode(StringInfo str,
 					 plan->startup_cost, plan->total_cost,
 					 ceil(plan->plan_rows / scaleFactor), plan->plan_width);
 
-	if (gp_resqueue_print_operator_memory_limits)
+	if (IsResGroupEnabled() ? gp_resgroup_print_operator_memory_limits
+							: gp_resqueue_print_operator_memory_limits)
```
Comment (guofengrichard) [inline src/backend/commands/explain.c]: haha:) Actually this is my first proposal regarding the change of the GUCs, but rejected by PM.

The concern is, changing the GUC in resource queue may also change the GUC value in resource group, and vice versa. It would confuse customer.

Comment (pivotal-ning-yu) [inline src/backend/commands/explain.c]: Hmm, then at least let's have a wrapper function similar to `GetMemoryPolicy()` to simplify the logic in the if block.

File: src/backend/utils/resource_manager/resource_manager.c [hunk 12]
```diff
@@ -33,3 +35,33 @@ IsResGroupEnabled(void)
...
+		gp_resmanager_memory_policy_auto_fixed_mem = gp_resqueue_memory_policy_auto_fixed_mem;
+	}
+	else if (IsResGroupEnabled() && !IsUnderPostmaster)
+	{
+		ResGroupControlInit();
+
+		gp_resmanager_memory_policy = gp_resgroup_memory_policy;
+		gp_log_resmanager_memory = gp_log_resgroup_memory;
+		gp_resmanager_print_operator_memory_limits = gp_resgroup_print_operator_memory_limits;
+		gp_resmanager_memory_policy_auto_fixed_mem = gp_resgroup_memory_policy_auto_fixed_mem;
```
Comment (pivotal-ning-yu) [inline src/backend/utils/resource_manager/resource_manager.c]: Will these GUCs be changed at runtime?

File: src/backend/utils/misc/guc_gp.c [hunk 13]
```diff
@@ -5347,6 +5380,15 @@ struct config_string ConfigureNamesString_gp[] =
 	},
 
 	{
+		{"gp_resgroup_memory_policy", PGC_SUSET, RESOURCES_MGM,
+			gettext_noop("Sets the policy for memory allocation of queries."),
+			gettext_noop("Valid values are AUTO, EAGER_FREE.")
```
Comment (pivotal-ning-yu) [inline src/backend/utils/misc/guc_gp.c]: Is NONE supported?

File: src/backend/utils/resource_manager/resource_manager.c (hunk 12 above)
Comment (pengzhout) [inline src/backend/utils/resource_manager/resource_manager.c]: Good catch, they are PGC_USERSET, so it can be changed at runtime, one option is change gp_resmanager* to gp_resmanager*_ptr ?

Comment (guofengrichard) [inline src/backend/utils/resource_manager/resource_manager.c]: Good catch! These GUCs can be changed inside session. We should change add assign function for each of them and change `gp_resmanager_*` in the assign function.

File: src/backend/utils/misc/guc_gp.c (hunk 13 above)
Comment (guofengrichard) [inline src/backend/utils/misc/guc_gp.c]: NONE is supported for resource group in now. 
Will remove it in later commit in this PR, or in a separate PR.

**Comment by pivotal-ning-yu** [comment]
LGTM

File: src/backend/cdb/cdbvars.c [hunk 14]
```diff
@@ -1311,11 +1310,53 @@ gpvars_show_gp_resqueue_memory_policy(void)
...
+		elog(ERROR, "unknown resource group memory policy: current policy is '%s'", gpvars_show_gp_resgroup_memory_policy());
+	else if (!pg_strcasecmp("auto", newval))
+		newtype = RESMANAGER_MEMORY_POLICY_AUTO;
+	else if (!pg_strcasecmp("eager_free", newval))
+		newtype = RESMANAGER_MEMORY_POLICY_EAGER_FREE;
+	else
+		elog(ERROR, "unknown resource group memory policy: current policy is '%s'", gpvars_show_gp_resgroup_memory_policy());
+
+	if (doit)
+	{
```
Comment (kainwen) [inline src/backend/cdb/cdbvars.c]: remove the {} for single line branch.

Comment (guofengrichard) [inline src/backend/cdb/cdbvars.c]: The `if (doit)` logic is common in GUC assign functions. To make these functions in the same style, I prefer to keep the `{}` here.

Comment (pivotal-ning-yu) [inline src/backend/cdb/cdbvars.c]: +1 on keeping existing style of this doit check.

File: src/backend/cdb/cdbvars.c [hunk 15]
```diff
@@ -1311,11 +1310,53 @@ gpvars_show_gp_resqueue_memory_policy(void)
...
+		case RESMANAGER_MEMORY_POLICY_NONE:
 			return "none";
-		case RESQUEUE_MEMORY_POLICY_AUTO:
+		case RESMANAGER_MEMORY_POLICY_AUTO:
 			return "auto";
-		case RESQUEUE_MEMORY_POLICY_EAGER_FREE:
+		case RESMANAGER_MEMORY_POLICY_EAGER_FREE:
+			return "eager_free";
+		default:
+			return "none";
```
Comment (danielgustafsson) [inline src/backend/cdb/cdbvars.c]: Shouldn't this trigger an `elog` for an internal error? Clearly we can't have any other value than none/auto/eager_free so hitting `default` here is a bug. It does come from existing code and wasn't introduced in this patch but we might as well address it.

Comment (guofengrichard) [inline src/backend/cdb/cdbvars.c]: Makes sense. An `elog` for `default` here is more reasonable than return `none`.
Will address this.
